### PR TITLE
Phong shading (again)

### PR DESCRIPTION
**Pull request by alexvidenov**

### CODE REVIEW

File: src/Marcher.hs [hunk 1]
```diff
@@ -139,24 +139,37 @@ type Direction = Vec3
...
-        Just newpos -> if equalWithinError epsilon 0 $ mag (light-newpos) then color else gray * color
-
-  where light = getSunPosition sett
-        newscene = mergeScenes s $ pointToScene light
-        epsilon = getTolerance sett
+    Just color -> color
 
 -- | Marches a ray through a scene until it hits an object. Returns Nothing if it goes outside the scene.
-rayMarch :: ImageSettings -> Scene -> Ray -> Maybe Position
+rayMarch :: ImageSettings -> Scene -> Ray -> Maybe Color
```
Comment (umnikos) [inline src/Marcher.hs]: Separating shading from pure marching was the entire point of having both `rayMarch` and `rayRender`

Comment (alexvidenov) [reply src/Marcher.hs]: Yeah, I just experimented with the shading and it was easier for me to put everything into one function.

Comment (alexvidenov) [reply src/Marcher.hs]: Anyway, I'll refactor it.

File: src/Marcher.hs [hunk 2]
```diff
@@ -137,28 +137,44 @@ type Direction = Vec3
...
-    Just pos -> let color = getColor $ snd $ s pos
-              in case rayMarch sett newscene (pos + 3*epsilon `scale` calcNormal sett s pos, normalize $ light - pos) of
-        Nothing -> color
-        Just newpos -> if equalWithinError epsilon 0 $ mag (light-newpos) then color else gray * color
-
-  where light = getSunPosition sett
-        newscene = mergeScenes s $ pointToScene light
-        epsilon = getTolerance sett
+    Just pos -> let (dist, material) = s pos
+                    end = getRenderDistance sett
```
Comment (umnikos) [inline src/Marcher.hs]: What do you need `end` for?

**Comment by umnikos** [comment]
Shouldn't `rayMarch.ppm` be in the `images/` folder?

**Comment by alexvidenov** [comment]
Oops, my bad. It should!

File: src/Marcher.hs [hunk 3]
```diff
@@ -137,28 +137,50 @@ type Direction = Vec3
...
+            lightDir = normalize $ pos-pt -- direction to the light source
+            -- diffused stuff
+            lambert = max 0 (normal `dot` lightDir )
+            diffused = lambert `scale` (color*materialColor)
+            -- specular stuff
+            r = reflect normal lightDir
+            specular = max 0 (glossPower * (max 0 $ r `dot` eye)**specPower) `scale` color
+            -- neither of which apply in shadow
+            shadow  | lambert == 0 = 0
+                    | otherwise = case rayMarch sett{getRenderDistance=mag (pos-pt)} s (pt + 0.0001 `scale` normal, lightDir) of
```
Comment (umnikos) [inline src/Marcher.hs]: Wait, what is that 0.0001 doing, anyway?

Comment (alexvidenov) [reply src/Marcher.hs]: This is basically the same arbitrary value we had in the previous version of marching to the sun to compute the shadow.

**Review by umnikos — APPROVED** [review]
Okay, I approve. Good work!

**Pull request merged**